The goal of this notebook is to produce connectomes and export connectome files for various exotic cases that will be run through the modularity algorithm. It uses the functions from get_connectome.py.

In [4]:
from neuprint import Client
# remove my token before making notebook public
c = Client('neuprint.janelia.org', dataset='hemibrain:v1.2.1', token='eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImdnMjExNEBjb2x1bWJpYS5lZHUiLCJsZXZlbCI6Im5vYXV0aCIsImltYWdlLXVybCI6Imh0dHBzOi8vbGgzLmdvb2dsZXVzZXJjb250ZW50LmNvbS9hLS9BT2gxNEdpb1lJLUVPLWdidGxPRTh6SmQ0eF9ZQ1Y4ZHF0YVFjWGlHeG5CMz1zOTYtYz9zej01MD9zej01MCIsImV4cCI6MTgxMDUyOTYzNH0.jv9eR0SH5RhfBdXrtp4r-dDFOhcsT8GBbE4v69ysCKs') 
c.fetch_version()

'0.1.0'

In [5]:
# import important stuff here
import numpy as np
import pandas as pd
import matplotlib

In [6]:
# body IDs of oviINs from Neuprint
oviINr_bodyID = 423101189
oviINl_bodyID = 485934965

# oviINr full connectome by cell type
This is oviINr's full connectome but the nodes are cell types and the edges are the collapsed weights by cell type. Nothing has been thresholded here so that we can compare to the full oviINr connectome by bodyId.

In [6]:
# use our function to get connectome by cell type
from get_connectome import get_connectome

oviINr_type_connectome = get_connectome(oviINr_bodyID, connectome_by_type=True)
oviINr_type_connectome

  0%|          | 0/23 [00:00<?, ?it/s]

type_pre    type_post  weight
0       5-HTPMPD01        AstA1       2
1       5-HTPMPD01        CL003       1
2       5-HTPMPD01      CL019_a       1
3       5-HTPMPD01        CL165       1
4       5-HTPMPD01        CL251       2
...            ...          ...     ...
212117       vpoEN         pC1c       2
212118       vpoEN         pC1d      14
212119       vpoEN         pC1e       2
212120       vpoEN  unspecified       4
212121       vpoEN        vpoDN      29

[212122 rows x 3 columns]

In [7]:
# make the connectome undirected
from get_connectome import connectome_to_undirected

oviINr_type_connectome_undirected = connectome_to_undirected(oviINr_type_connectome)
oviINr_type_connectome_undirected

source   target  weight
0       5-HTPMPD01    AstA1       2
1       5-HTPMPD01    CL003       3
2       5-HTPMPD01  CL019_a       1
3       5-HTPMPD01    CL165       1
4       5-HTPMPD01    CL251       2
...            ...      ...     ...
163201       vpoEN    oviIN       4
163202       vpoEN     pC1b       1
163203       vpoEN     pC1c       2
163204       vpoEN     pC1d      14
163205       vpoEN     pC1e       2

[163206 rows x 3 columns]

As of now, the C code won't work with the connectome by types data. Rhessa has some code to assign 8-digit numbers to cell types as a workaround. 
**Update:** I fixed this issue. See test_analysis_pipeline.ipynb for details.

In [8]:
import os

# file path 
os.chdir('/Users/ggutierr/My Drive (ggutierr@barnard.edu)/GitHub/oviIN-analyses-gabrielle/')
path = os.getcwd()

In [9]:
# export the connectome to a csv file
#oviINr_type_connectome_undirected.to_csv('for_running_modularity/oviINr_type/oviINr_type_connectome_undirected.txt', index = False, header=False)
oviINr_type_connectome_undirected.to_csv('oviINr_type_connectome_undirected.txt', index = False, header=False)

# oviINr full connectome with left instances removed
This one is tricky because we can't use the functions we've made for getting connectomes. 

In [17]:
from neuprint import fetch_adjacencies

# the 1st df returns all the neurons involved in making the specified connections
pre, pre_conns = fetch_adjacencies(None, oviINr_bodyID)
post, post_conns = fetch_adjacencies(oviINr_bodyID, None)

In [18]:
pre

bodyId       type         instance
0      423101189      oviIN          oviIN_R
1      234630133     SMP184  SMP184(PDL05)_L
2      263674097  LHPD2a5_a      LHPD2a5_a_R
3      266187480     SMP349         SMP349_R
4      266187559     SLP399         SLP399_R
...          ...        ...              ...
2520  5901231318       None             None
2521  5901232053     SMP272  SMP272(PDL21)_L
2522  6400000773     SMP411         SMP411_R
2523  7112622044     LAL137  LAL137(PVL05)_L
2524  7112624972       None             None

[2525 rows x 3 columns]

Get rid of the left instances now before we concatenate pre and post bodyIds. This is done by simply removing the rows of the pre and post dataframes where the instance ends with _L.

In [39]:
# replace None with 'unspecified'
pre.fillna('unspecified', inplace=True)
post.fillna('unspecified', inplace=True)

# remove rows where instance has _L at the end of instance
post = post[~post.instance.str.endswith('_L')]
pre = pre[~pre.instance.str.endswith('_L')]
pre

bodyId         type     instance
0      423101189        oviIN      oviIN_R
2      263674097    LHPD2a5_a  LHPD2a5_a_R
3      266187480       SMP349     SMP349_R
4      266187559       SLP399     SLP399_R
5      267214250         pC1b       pC1b_R
...          ...          ...          ...
2518  5901227238  unspecified  unspecified
2519  5901230025  unspecified  unspecified
2520  5901231318  unspecified  unspecified
2522  6400000773       SMP411     SMP411_R
2524  7112624972  unspecified  unspecified

[2087 rows x 3 columns]

In [40]:
# combine unique pre and post bodyIds
partners = pd.concat([pre['bodyId'], post['bodyId']]).unique()

# turn it back into a series
partners = pd.Series(partners)

The concatenated bodyIds are then used to fetch_adjacencies among all of these non-left instances in oviINr's connectome. And of course we collapse the weights across the ROIs.

In [41]:
# get connections among neurons using the bodyIds from partners
partners_, connectome = fetch_adjacencies(partners, partners)

# get rid of the ROI column and group bodyId_pre and bodyId_post by summing weights across ROIs
connectome = connectome.groupby(['bodyId_pre', 'bodyId_post'], as_index=False)['weight'].sum()

  0%|          | 0/20 [00:00<?, ?it/s]

In [42]:
connectome

bodyId_pre  bodyId_post  weight
0        203594164    203598466       3
1        203594164    296358430       1
2        203594164    298944745       1
3        203594164    326530038       2
4        203594164    328939480       1
...            ...          ...     ...
336596  7112624972    610312734       1
336597  7112624972    732034854       1
336598  7112625115    643393688       1
336599  7112625115   5813055749       1
336600  7112625282    769377106       1

[336601 rows x 3 columns]

Now we can pass it into the connectome_to_undirected function before saving.

In [43]:
from get_connectome import connectome_to_undirected

oviINr_Lremoved_connectome_undirected = connectome_to_undirected(connectome)
oviINr_Lremoved_connectome_undirected

source      target  weight
0        203594164   203598466      94
1        203594164   296358430       1
2        203594164   298944745       2
3        203594164   326530038       4
4        203594164   328939480       1
...            ...         ...     ...
284529  7112624972   610312734       1
284530  7112624972   732034854       1
284531  7112625115   643393688       1
284532  7112625115  5813055749       1
284533  7112625282   769377106       1

[284534 rows x 3 columns]

In [44]:
# export the connectome to a csv file
#oviINr_Lremoved_connectome_undirected.to_csv('for_running_modularity/oviINr_left_removed/oviINr_Lremoved_connectome_undirected.txt', index = False, header=False)
oviINr_Lremoved_connectome_undirected.to_csv('oviINr_Lremoved_connectome_undirected.txt', index = False, header=False)

# oviINr full connectome with thresholded edges
The thresholding happens on the connectome edges. I fully expect the modularity results on this to look a lot like the Sankeys I made for prominent partners and I expect the modularity to give results at the coarse resolution that are very similar to not thresholding.

In [33]:
# use our function to get connectome by cell type
from get_connectome import get_connectome

weight_th = 20
oviINr_connectome_thresh = get_connectome(oviINr_bodyID, weight_threshold=weight_th)
oviINr_connectome_thresh

  0%|          | 0/23 [00:00<?, ?it/s]

bodyId_pre  bodyId_post  weight
4        203253253    296838079      35
5        203253253    296855409      20
12       203253253    326530038      31
73       203253253   5813020684      33
74       203253253   5813020696      64
...            ...          ...     ...
537157  7112622044   5813058319      49
537182  7112622044   5813087891      29
537184  7112622044   5813089157      21
537192  7112622044   5813132515      52
537195  7112622044   5901221584      74

[19514 rows x 3 columns]

In [34]:
# make the connectome undirected
from get_connectome import connectome_to_undirected

oviINr_connectome_thresh_undirected = connectome_to_undirected(oviINr_connectome_thresh)
oviINr_connectome_thresh_undirected

source      target  weight
0       203253253   296838079      35
1       203253253   296855409      20
2       203253253   326530038      67
3       203253253  5813020684      33
4       203253253  5813020696      64
...           ...         ...     ...
18522  7112622044  5813021602      32
18523  7112622044  5813058319      49
18524  7112622044  5813087891      29
18525  7112622044  5813089157      21
18526  7112622044  5813132515      52

[18527 rows x 3 columns]

In [36]:
# export the connectome to a csv file
oviINr_connectome_thresh_undirected.to_csv('oviINr_connectome_thresh20_undirected.txt', index = False, header=False)

# oviINr full connectome with thresholded edges before assembling connectome
In other words, the thresholding happens for the connections between oviINr and its partners. This filters out the weak partners of oviIN and leaves its strong partners to be a part of the connectome.

In [45]:
# inputs to oviINr
from neuprint import fetch_simple_connections
ovi_inputs = fetch_simple_connections(None,oviINr_bodyID)
ovi_inputs

bodyId_pre  bodyId_post  weight type_pre type_post instance_pre  \
0      611015907    423101189     287   SMP176     oviIN     SMP176_R   
1     5813087891    423101189     255   SMP544     oviIN     SMP544_R   
2     1598011458    423101189     243   LAL134     oviIN     LAL134_R   
3     5813022375    423101189     236    IB017     oviIN      IB017_R   
4      448342846    423101189     225   CRE075     oviIN     CRE075_R   
...          ...          ...     ...      ...       ...          ...   
2519  5901221890    423101189       1   SLP422     oviIN     SLP422_R   
2520  5901226036    423101189       1     None     oviIN         None   
2521  5901230025    423101189       1     None     oviIN         None   
2522  5901231318    423101189       1     None     oviIN         None   
2523  7112624972    423101189       1     None     oviIN         None   

     instance_post                                       conn_roiInfo  
0          oviIN_R  {'INP': {'pre': 54, 'post': 53}, 'CRE(R)': {'p...  
1          oviIN_R  {'LX(R)': {'pre': 66, 'post': 65}, 'LAL(R)': {...  
2          oviIN_R  {'LX(R)': {'pre': 42, 'post': 42}, 'LAL(R)': {...  
3          oviIN_R  {'SNP(R)': {'pre': 97, 'post': 97}, 'SMP(R)': ...  
4          oviIN_R  {'INP': {'pre': 199, 'post': 199}, 'CRE(R)': {...  
...            ...                                                ...  
2519       oviIN_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
2520       oviIN_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
2521       oviIN_R  {'VMNP': {'pre': 1, 'post': 1}, 'VES(R)': {'pr...  
2522       oviIN_R  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  
2523       oviIN_R  {'INP': {'pre': 1, 'post': 1}, 'CRE(R)': {'pre...  

[2524 rows x 8 columns]

In [46]:
# outputs to oviINr
from neuprint import fetch_simple_connections
ovi_outputs = fetch_simple_connections(oviINr_bodyID,None)
ovi_outputs

bodyId_pre  bodyId_post  weight type_pre type_post instance_pre  \
0      423101189   5813021602     320    oviIN    SMP198      oviIN_R   
1      423101189    452689494     239    oviIN    SMP550      oviIN_R   
2      423101189    611629428     238    oviIN    CRE077      oviIN_R   
3      423101189    542376096     208    oviIN    SMP237      oviIN_R   
4      423101189   5813021005     199    oviIN    SMP507      oviIN_R   
...          ...          ...     ...      ...       ...          ...   
3436   423101189   5901197274       1    oviIN    SMP528      oviIN_R   
3437   423101189   5901203310       1    oviIN      None      oviIN_R   
3438   423101189   5901220525       1    oviIN      None      oviIN_R   
3439   423101189   7112625063       1    oviIN      None      oviIN_R   
3440   423101189   7112625282       1    oviIN      None      oviIN_R   

        instance_post                                       conn_roiInfo  
0            SMP198_R  {'SNP(R)': {'pre': 276, 'post': 276}, 'SMP(R)'...  
1            SMP550_R  {'SNP(R)': {'pre': 239, 'post': 239}, 'SMP(R)'...  
2            CRE077_R  {'SNP(R)': {'pre': 216, 'post': 216}, 'SMP(R)'...  
3            SMP237_R  {'SNP(R)': {'pre': 203, 'post': 203}, 'SMP(R)'...  
4            SMP507_R  {'SNP(R)': {'pre': 199, 'post': 199}, 'SMP(R)'...  
...               ...                                                ...  
3436  SMP528(PDM24)_L  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  
3437        (PVM12)_L  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  
3438             None  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
3439             None  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
3440             None  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  

[3441 rows x 8 columns]

In [47]:
weight_th = 20

# threshold ovi_inputs weight
ovi_inputs = ovi_inputs[ovi_inputs.weight >= weight_th]

# threshold ovi_outputs weight
ovi_outputs = ovi_outputs[ovi_outputs.weight >= weight_th]

In [48]:
ovi_outputs

bodyId_pre  bodyId_post  weight type_pre type_post instance_pre  \
0     423101189   5813021602     320    oviIN    SMP198      oviIN_R   
1     423101189    452689494     239    oviIN    SMP550      oviIN_R   
2     423101189    611629428     238    oviIN    CRE077      oviIN_R   
3     423101189    542376096     208    oviIN    SMP237      oviIN_R   
4     423101189   5813021005     199    oviIN    SMP507      oviIN_R   
..          ...          ...     ...      ...       ...          ...   
345   423101189    800622401      20    oviIN     PAM13      oviIN_R   
346   423101189    822078542      20    oviIN    SMP381      oviIN_R   
347   423101189    894172798      20    oviIN      None      oviIN_R   
348   423101189   5813069072      20    oviIN    SMP287      oviIN_R   
349   423101189   5813081474      20    oviIN      None      oviIN_R   

      instance_post                                       conn_roiInfo  
0          SMP198_R  {'SNP(R)': {'pre': 276, 'post': 276}, 'SMP(R)'...  
1          SMP550_R  {'SNP(R)': {'pre': 239, 'post': 239}, 'SMP(R)'...  
2          CRE077_R  {'SNP(R)': {'pre': 216, 'post': 216}, 'SMP(R)'...  
3          SMP237_R  {'SNP(R)': {'pre': 203, 'post': 203}, 'SMP(R)'...  
4          SMP507_R  {'SNP(R)': {'pre': 199, 'post': 199}, 'SMP(R)'...  
..              ...                                                ...  
345  PAM13(B'1ap)_R  {'SNP(R)': {'pre': 20, 'post': 20}, 'SMP(R)': ...  
346        SMP381_R  {'INP': {'pre': 13, 'post': 13}, 'CRE(R)': {'p...  
347            None  {'SNP(R)': {'pre': 18, 'post': 18}, 'SMP(R)': ...  
348        SMP287_R  {'SNP(R)': {'pre': 20, 'post': 20}, 'SMP(R)': ...  
349     (MBDLaxon1)  {'SNP(R)': {'pre': 15, 'post': 15}, 'SMP(R)': ...  

[350 rows x 8 columns]

In [49]:
from neuprint import fetch_adjacencies

pre, pre_conns = fetch_adjacencies(ovi_inputs['bodyId_pre'], oviINr_bodyID)
post, post_conns = fetch_adjacencies(oviINr_bodyID, ovi_outputs['bodyId_post'])

In [50]:
post

bodyId    type          instance
0     423101189   oviIN           oviIN_R
1     234630133  SMP184   SMP184(PDL05)_L
2     266187383  SMP347          SMP347_R
3     266187559  SLP399          SLP399_R
4     267551639    pC1c            pC1c_R
..          ...     ...               ...
346  5901203505  SLP212          SLP212_R
347  5901212906    None  SLPXXX(SCBXXX)_R
348  5901213440  CRE011          CRE011_R
349  5901213624  SMP405          SMP405_R
350  5901232053  SMP272   SMP272(PDL21)_L

[351 rows x 3 columns]

In [51]:
# combine unique pre and post bodyIds
partners = pd.concat([pre['bodyId'], post['bodyId']]).unique()
# turn it back into a series
partners = pd.Series(partners)

# get connections among neurons using the bodyIds from partners
partners_, connectome = fetch_adjacencies(partners, partners)

  0%|          | 0/3 [00:00<?, ?it/s]

In [52]:
# get rid of the ROI column and group bodyId_pre and bodyId_post by summing weights across ROIs
connectome = connectome.groupby(['bodyId_pre', 'bodyId_post'], as_index=False)['weight'].sum()

In [53]:
connectome

bodyId_pre  bodyId_post  weight
0       234630133    296838079       1
1       234630133    296855409       4
2       234630133    298603730       1
3       234630133    328244591       1
4       234630133    328943204      21
...           ...          ...     ...
31917  5901232053   5813055949      19
31918  5901232053   5813056887       2
31919  5901232053   5813069072       2
31920  5901232053   5813132515       4
31921  5901232053   5901213624       1

[31922 rows x 3 columns]

In [54]:
# make the connectome undirected
from get_connectome import connectome_to_undirected

oviINr_thresh_connectome_undirected = connectome_to_undirected(connectome)
oviINr_thresh_connectome_undirected

source      target  weight
0       234630133   296838079       1
1       234630133   296855409       4
2       234630133   298603730       1
3       234630133   328244591       1
4       234630133   328943204      29
...           ...         ...     ...
24975  5901232053  5813055949      19
24976  5901232053  5813056887       2
24977  5901232053  5813069072       2
24978  5901232053  5813132515       4
24979  5901232053  5901213624       1

[24980 rows x 3 columns]

In [55]:
# export the connectome to a csv file
oviINr_thresh_connectome_undirected.to_csv('oviINr_thresh20_connectome_undirected.txt', index = False, header=False)

# oviENr full connectome
There is 1 oviEN in each hemisphere and it is currently annotated as SMP550.

In [56]:
from neuprint import fetch_neurons
from neuprint import NeuronCriteria as NC

# the oviENr
oviENr, roi_counts_df = fetch_neurons(452689494)
oviENr

bodyId  instance    type   pre  post  downstream  upstream  mito  \
0  452689494  SMP550_R  SMP550  1414  4831       12881      4831   555   

         size  status  cropped     statusLabel cellBodyFiber  somaRadius  \
0  2052918412  Traced    False  Roughly traced         PDM29         347   

           somaLocation                                            roiInfo  \
0  [19511, 18367, 6016]  {'SNP(R)': {'pre': 1248, 'post': 3275, 'downst...   

             notes                                          inputRois  \
0  pSP7 candidates  [AVLP(R), INP, LH(R), PLP(R), SCL(R), SIP(R), ...   

                                          outputRois  
0  [AVLP(R), INP, LH(R), PLP(R), SCL(R), SIP(R), ...

In [61]:
# use our function to get connectome
from get_connectome import get_connectome

oviENr_connectome = get_connectome(oviENr.bodyId[0])
oviENr_connectome

  0%|          | 0/5 [00:00<?, ?it/s]

bodyId_pre  bodyId_post  weight
0       263674097    298258513      28
1       263674097    329225149       2
2       263674097    361312808       6
3       263674097    422751191      47
4       263674097    423101189       2
...           ...          ...     ...
59900  5901232295    671268907       1
59901  5901232295   1042656011       1
59902  5901232295   5813009996       2
59903  5901232295   5813010210       1
59904  5901232295   5813050455       1

[59905 rows x 3 columns]

In [62]:
# make the connectome undirected
from get_connectome import connectome_to_undirected

oviENr_connectome_undirected = connectome_to_undirected(oviENr_connectome)
oviENr_connectome_undirected

source      target  weight
0       263674097   298258513      32
1       263674097   329225149       2
2       263674097   361312808       6
3       263674097   422751191      47
4       263674097   423101189       8
...           ...         ...     ...
48850  5901232295   640596169       1
48851  5901232295  1042656011       1
48852  5901232295  5813009996       2
48853  5901232295  5813010210       1
48854  5901232295  5813050455       1

[48855 rows x 3 columns]

In [63]:
# export the connectome to a csv file
oviENr_connectome_undirected.to_csv('oviENr_connectome_undirected.txt', index = False, header=False)

# oviINr full connectome with oviINr removed
I will remove just oviINr from its connectome. The idea is that since we know that oviINr is connected to everything in its connectome that maybe this is having some weird effect on the results. It may also be worth investigating what happens when both oviINs are removed.

In [64]:
# use our function to get connectome
from get_connectome import get_connectome

oviINr_selfremoved_connectome = get_connectome(oviINr_bodyID,exclude_main_neurons=True)
oviINr_selfremoved_connectome

  0%|          | 0/23 [00:00<?, ?it/s]

bodyId_pre  bodyId_post  weight
0        203253253    203594164       2
1        203253253    295802733       1
2        203253253    296194535       1
3        203253253    296509709       7
4        203253253    296838079      35
...            ...          ...     ...
531246  7112625282    674882250       1
531247  7112625282    769377106       1
531248  7112625282    800373114       1
531249  7112625282   1163199311       1
531250  7112625282   5813062838       1

[531251 rows x 3 columns]

In [65]:
# make the connectome undirected
from get_connectome import connectome_to_undirected

oviINr_selfremoved_connectome_undirected = connectome_to_undirected(oviINr_selfremoved_connectome)
oviINr_selfremoved_connectome_undirected

source      target  weight
0        203253253   203594164       5
1        203253253   295802733       1
2        203253253   296194535       1
3        203253253   296509709      26
4        203253253   296838079      35
...            ...         ...     ...
448074  7112625282   674882250       1
448075  7112625282   769377106       1
448076  7112625282   800373114       1
448077  7112625282  1163199311       1
448078  7112625282  5813062838       1

[448079 rows x 3 columns]

In [66]:
# export the connectome to a csv file
oviINr_selfremoved_connectome_undirected.to_csv('oviINr_selfremoved_connectome_undirected.txt', index = False, header=False)

# oviINr full connectome: multiple runs with different random seeds
This is the classic oviINr full connectome but the modularity algorithm will be given a new random seed each time we run it. The same set of files generated here will be used for those multiple runs. The goal of doing this is to determine how much variation the algorithm gives us.


In [67]:
# use our function to get connectome
from get_connectome import get_connectome

oviINr_connectome = get_connectome(oviINr_bodyID)
oviINr_connectome

  0%|          | 0/23 [00:00<?, ?it/s]

bodyId_pre  bodyId_post  weight
0        203253253    203594164       2
1        203253253    295802733       1
2        203253253    296194535       1
3        203253253    296509709       7
4        203253253    296838079      35
...            ...          ...     ...
537211  7112625282    674882250       1
537212  7112625282    769377106       1
537213  7112625282    800373114       1
537214  7112625282   1163199311       1
537215  7112625282   5813062838       1

[537216 rows x 3 columns]

In [68]:
# make the connectome undirected
from get_connectome import connectome_to_undirected

oviINr_connectome_undirected = connectome_to_undirected(oviINr_connectome)
oviINr_connectome_undirected

source      target  weight
0        203253253   203594164       5
1        203253253   295802733       1
2        203253253   296194535       1
3        203253253   296509709      26
4        203253253   296838079      35
...            ...         ...     ...
452622  7112625282   674882250       1
452623  7112625282   769377106       1
452624  7112625282   800373114       1
452625  7112625282  1163199311       1
452626  7112625282  5813062838       1

[452627 rows x 3 columns]

In [69]:
# export the connectome to a csv file
oviINr_connectome_undirected.to_csv('oviINr_connectome_undirected.txt', index = False, header=False)

Below I generate a bunch of random seeds that can be used for each run of the same data.

In [71]:
# generate 10 random numbers 5-digit numbers
import random
randomseeds = []
for i in range(0,10):
    n = random.randint(10000,99999)
    randomseeds.append(n)
randomseeds

[63039, 62833, 93507, 36474, 37632, 46807, 66667, 16104, 75525, 46418]

In [72]:
# export them to a text file
with open("randomseeds.txt", "w") as output:
    for row in randomseeds:
        output.write(str(row) + '\n')

# oviINr inputs connectome
This is the default inputs connectome for oviINr that keeps oviINr in the network. 

In [14]:
# use our function to get connectome
from get_connectome import get_connectome

oviINr_connectome = get_connectome(oviINr_bodyID, connectome_scope='input')
oviINr_connectome

  0%|          | 0/13 [00:00<?, ?it/s]

bodyId_pre  bodyId_post  weight
0        234630133    296855409       4
1        234630133    296885362       2
2        234630133    297908595       1
3        234630133    298603730       1
4        234630133    326253554       1
...            ...          ...     ...
245366  7112622044   5813132515      52
245367  7112624972    423101189       1
245368  7112624972    576566223       1
245369  7112624972    610312734       1
245370  7112624972    827893839       1

[245371 rows x 3 columns]

FYI, oviINr is in the input connectome by default because when we fetch_adjacencies to get the pre neurons, the first dataframe returns all the pre neurons as well as the queried neurons. 

In [31]:
# checking whether oviINr is in the input connectome by default
oviINr_connectome[oviINr_connectome['bodyId_post'] == oviINr_bodyID]

bodyId_pre  bodyId_post  weight
27       234630133    423101189       2
179      263674097    423101189       2
273      266187480    423101189       1
319      266187559    423101189       3
457      267214250    423101189       9
...            ...          ...     ...
244783  5901231318    423101189       1
244859  5901232053    423101189       3
245040  6400000773    423101189       2
245128  7112622044    423101189       3
245367  7112624972    423101189       1

[2524 rows x 3 columns]

In [32]:
# make the connectome undirected
from get_connectome import connectome_to_undirected

oviINr_connectome_undirected = connectome_to_undirected(oviINr_connectome)
oviINr_connectome_undirected

source      target  weight
0        234630133   296855409       4
1        234630133   296885362       2
2        234630133   297908595       1
3        234630133   298603730       1
4        234630133   326253554      41
...            ...         ...     ...
199656  7112622044  5813080768      10
199657  7112622044  5813096198       2
199658  7112622044  5813098223       5
199659  7112624972   576566223       1
199660  7112624972   610312734       1

[199661 rows x 3 columns]

In [6]:
# export the connectome to a txt file
oviINr_connectome_undirected.to_csv('for_running_modularity/oviINr_inputs/oviINr_in_connectome_undirected.txt', index = False, header=False)

In [33]:
test1 = oviINr_connectome
test2 = oviINr_connectome_undirected

# oviINr input connectome: with oviINr removed
Inputs to oviINr only with oviINr excluded from the network.

In [29]:
# use our function to get connectome
from get_connectome import get_connectome

oviINr_connectome = get_connectome(oviINr_bodyID, exclude_main_neurons=True, connectome_scope='input')
oviINr_connectome

  0%|          | 0/13 [00:00<?, ?it/s]

bodyId_pre  bodyId_post  weight
0        234630133    296855409       4
1        234630133    296885362       2
2        234630133    297908595       1
3        234630133    298603730       1
4        234630133    326253554       1
...            ...          ...     ...
241425  7112622044   5813111272       1
241426  7112622044   5813132515      52
241427  7112624972    576566223       1
241428  7112624972    610312734       1
241429  7112624972    827893839       1

[241430 rows x 3 columns]

In [82]:
# export the connectome to a txt file
oviINr_connectome.to_csv('for_running_modularity/oviINr_inputs_selfremoved/oviINr_in_connectome_selfremoved.txt', index = False, header=False)

In [30]:
# make the connectome undirected
from get_connectome import connectome_to_undirected

oviINr_connectome_undirected = connectome_to_undirected(oviINr_connectome)
oviINr_connectome_undirected

source      target  weight
0        234630133   296855409       4
1        234630133   296885362       2
2        234630133   297908595       1
3        234630133   298603730       1
4        234630133   326253554      41
...            ...         ...     ...
197132  7112622044  5813080768      10
197133  7112622044  5813096198       2
197134  7112622044  5813098223       5
197135  7112624972   576566223       1
197136  7112624972   610312734       1

[197137 rows x 3 columns]

In [6]:
# export the connectome to a txt file
oviINr_connectome_undirected.to_csv('for_running_modularity/oviINr_inputs_selfremoved/oviINr_in_connectome_undirected_selfremoved.txt', index = False, header=False)

In [6]:
oviINr_connectome['bodyId_pre'].nunique()

2494

# Test of the sub-connectome composition
If we use fetch_adjacencies to grab the neurons that are inputs to oviIN, what I show below is that by default it returns neurons of various status. Most are Traced but several are labeled 'Assign' and 'Orphan'. Among Traced neurons are some that are incomplete because they are labeled 'Leaves' or 'Roughly Traced' in the statusLabel column. 

In [26]:
from neuprint import fetch_adjacencies
pre, pre_conns = fetch_adjacencies(None, oviINr_bodyID)
pre

bodyId       type         instance
0      423101189      oviIN          oviIN_R
1      234630133     SMP184  SMP184(PDL05)_L
2      263674097  LHPD2a5_a      LHPD2a5_a_R
3      266187480     SMP349         SMP349_R
4      266187559     SLP399         SLP399_R
...          ...        ...              ...
2520  5901231318       None             None
2521  5901232053     SMP272  SMP272(PDL21)_L
2522  6400000773     SMP411         SMP411_R
2523  7112622044     LAL137  LAL137(PVL05)_L
2524  7112624972       None             None

[2525 rows x 3 columns]

In [27]:
# get these neurons so that we can see if they are Traced
from neuprint import fetch_neurons
pre_neurons,_ = fetch_neurons(pre['bodyId'].unique())
pre_neurons

bodyId         instance       type   pre  post  downstream  \
0      234630133  SMP184(PDL05)_L     SMP184  1075  2470        7782   
1      263674097      LHPD2a5_a_R  LHPD2a5_a   133   547        1070   
2      266187480         SMP349_R     SMP349   190   854        1519   
3      266187559         SLP399_R     SLP399   216   749        1794   
4      267214250           pC1b_R       pC1b   816  3187        5440   
...          ...              ...        ...   ...   ...         ...   
2520  5901231318             None       None    55    66         542   
2521  5901232053  SMP272(PDL21)_L     SMP272  1064  1640        7150   
2522  6400000773         SMP411_R     SMP411   182   522        1545   
2523  7112622044  LAL137(PVL05)_L     LAL137  1376  2087       11635   
2524  7112624972             None       None     4    42          19   

      upstream  mito        size  status cropped     statusLabel  \
0         2470   514  1650479006  Traced    True          Leaves   
1          547   146   408560985  Traced   False  Roughly traced   
2          854   165   563941715  Traced   False          Traced   
3          749   125   539797068  Traced   False  Roughly traced   
4         3187   654  3805489752  Traced   False          Traced   
...        ...   ...         ...     ...     ...             ...   
2520        66    23    61516795  Traced    True          Leaves   
2521      1640   358  1421485085  Traced   False  Roughly traced   
2522       522   135   503262274  Traced   False  Roughly traced   
2523      2087   534  1529252938  Traced   False  Roughly traced   
2524        42     5     8531939  Assign    None       0.5assign   

     cellBodyFiber  somaRadius           somaLocation  \
0             None         NaN                   None   
1            PDL06       268.5    [5386, 20096, 4080]   
2            PDM07       238.5   [18808, 27714, 4256]   
3            PDM07       290.5   [17838, 26568, 3924]   
4            PDM09       446.5  [18931, 10896, 14728]   
...            ...         ...                    ...   
2520          None         NaN                   None   
2521          None         NaN                   None   
2522         PDM09       321.5  [22433, 11755, 18464]   
2523          None         NaN                   None   
2524          None         NaN                   None   

                                                roiInfo              notes  \
0     {'SNP(L)': {'pre': 153, 'post': 1758, 'downstr...               None   
1     {'SNP(R)': {'pre': 121, 'post': 330, 'downstre...               None   
2     {'SNP(R)': {'pre': 190, 'post': 854, 'downstre...               None   
3     {'SNP(R)': {'pre': 216, 'post': 744, 'downstre...               None   
4     {'SNP(R)': {'pre': 547, 'post': 2474, 'downstr...               None   
...                                                 ...                ...   
2520  {'SNP(L)': {'pre': 55, 'post': 66, 'downstream...               None   
2521  {'SNP(L)': {'pre': 464, 'post': 967, 'downstre...               None   
2522  {'SNP(R)': {'pre': 180, 'post': 467, 'downstre...               None   
2523  {'SNP(R)': {'pre': 304, 'post': 130, 'downstre...  VeLC-(LC)(c)Sm(b)   
2524  {'SNP(R)': {'pre': 1, 'post': 6, 'downstream':...               None   

                                              inputRois  \
0     [CRE(-ROB,-RUB)(R), CRE(R), INP, LAL(-GA)(R), ...   
1     [CRE(-ROB,-RUB)(R), CRE(R), INP, LH(R), PLP(R)...   
2                      [SIP(R), SLP(R), SMP(R), SNP(R)]   
3                       [LH(R), SLP(R), SMP(R), SNP(R)]   
4     [AOTU(R), AVLP(R), ICL(R), INP, SCL(R), SIP(R)...   
...                                                 ...   
2520                                   [SMP(L), SNP(L)]   
2521  [CRE(-RUB)(L), CRE(L), INP, LAL(L), LX(L), RUB...   
2522  [INP, LH(R), MB(+ACA)(R), PLP(R), SCL(R), SIP(...   
2523  [CRE(-ROB,-RUB)(R), CRE(-RUB)(L), CRE(L), CRE(...   
2524   [CRE(-ROB,-RUB)(R), CRE(R), INP, SMP(R), SNP(R)]  

In [8]:
# get counts of status
pre_neurons['status'].value_counts()

status
Traced    2316
Assign     134
Orphan      75
Name: count, dtype: int64

In [17]:
pre_neurons[pre_neurons['status'] == 'Traced']['statusLabel'].value_counts()

statusLabel
Roughly traced    1753
Leaves             483
Traced              80
Name: count, dtype: int64

In [9]:
pre_neurons[pre_neurons['status'] == 'Assign']

bodyId instance  type  pre  post  downstream  upstream  mito  \
585    581336621     None  None    1    11          10        11     1   
710    613744224     None  None    3     0          11         0     1   
760    642041782     None  None    2     7           7         7     0   
779    643160749     None  None    2     7           6         7     2   
792    643864734     None  None    1    12           7        12     2   
...          ...      ...   ...  ...   ...         ...       ...   ...   
2149  2347274239     None  None    3     1          15         1     0   
2151  2377972598     None  None    2     0           9         0     1   
2478  5813100373     None  None    3     1          16         1     0   
2517  5901226036     None  None    2     2          18         2     0   
2524  7112624972     None  None    4    42          19        42     5   

         size  status cropped statusLabel cellBodyFiber  somaRadius  \
585   1984795  Assign    None   0.5assign          None         NaN   
710   1053477  Assign    None   0.5assign          None         NaN   
760   1500538  Assign    None   0.5assign          None         NaN   
779   1062206  Assign    None   0.5assign          None         NaN   
792   2163539  Assign    None   0.5assign          None         NaN   
...       ...     ...     ...         ...           ...         ...   
2149   605561  Assign    None   0.5assign          None         NaN   
2151   666418  Assign    None   0.5assign          None         NaN   
2478  1657209  Assign    None   0.5assign          None         NaN   
2517   104391  Assign    None   0.5assign          None         NaN   
2524  8531939  Assign    None   0.5assign          None         NaN   

     somaLocation                                            roiInfo notes  \
585          None  {'SNP(R)': {'pre': 1, 'post': 11, 'downstream'...  None   
710          None  {'SNP(R)': {'pre': 3, 'downstream': 11, 'mito'...  None   
760          None  {'SNP(R)': {'pre': 2, 'post': 7, 'downstream':...  None   
779          None  {'SNP(L)': {'pre': 2, 'post': 7, 'downstream':...  None   
792          None  {'SNP(L)': {'pre': 1, 'post': 12, 'downstream'...  None   
...           ...                                                ...   ...   
2149         None  {'VMNP': {'pre': 3, 'post': 1, 'downstream': 1...  None   
2151         None  {'VMNP': {'pre': 2, 'downstream': 9, 'mito': 1...  None   
2478         None  {'SNP(L)': {'pre': 3, 'post': 1, 'downstream':...  None   
2517         None  {'SNP(R)': {'pre': 2, 'post': 2, 'downstream':...  None   
2524         None  {'SNP(R)': {'pre': 1, 'post': 6, 'downstream':...  None   

                                             inputRois  \
585                                   [SMP(R), SNP(R)]   
710                                                 []   
760                                   [SMP(R), SNP(R)]   
779                                   [SMP(L), SNP(L)]   
792                                   [SMP(L), SNP(L)]   
...                                                ...   
2149                                    [VES(R), VMNP]   
2151                                                []   
2478                                  [SMP(L), SNP(L)]   
2517                                  [SMP(R), SNP(R)]   
2524  [CRE(-ROB,-RUB)(R), CRE(R), INP, SMP(R), SNP(R)]   

                                            outputRois  
585                                   [SMP(R), SNP(R)]  
710                                   [SMP(R), SNP(R)]  
760                                   [SMP(R), SNP(R)]  
779                                   [SMP(L), SNP(L)]  
792                                   [SMP(L), SNP(L)]  
...                                                ...  
2149                                    [VES(R), VMNP]  
2151                                    [VES(R), VMNP]  
2478                                  [SMP(L), SNP(L)]  
2517                                  [SMP(R), SNP(R)]  
2524  [

In [18]:
pre_neurons[pre_neurons['status'] == 'Assign']['instance'].value_counts()

Series([], Name: count, dtype: int64)

## Comparison of sub-connectome datasets
Rhessa says that the discrepancy between her data and Pranjal's has to do with Janelia changing the weights of the connectome since last year.

This is Rhessa's modularity results for the oviIN input connectome from August 22, 2023 which is over a year old. There are only 2511 neurons which is 14 fewer than the 2525 returned by fetch_adjacencies, but that discrepancy could be explained if those are the 14 neurons that don't connect to anything else in oviIN's inputs. 

In [19]:
# load Rhessa's data from August 22, 2023
ovi_prepro = pd.read_csv('ovi_preprocessed/preprocessed_inputs-v1.2.1/preprocessed_nodes.csv')#, index_col=0)
ovi_prepro

id   key  0.0  0.05  0.1  0.5  0.75  1.0   instance celltype  \
0     1003215282     1    1     1    1    1     1    1    CL229_R    CL229   
1     1005952640     2    2     1    1    2     2    2    IB058_R    IB058   
2     1006928515     3    1     1    1    3     3    3    CL300_R    CL300   
3     1007260806     4    2     1    1    4     4    4    CL301_R    CL301   
4     1008024276     5    3     2    2    5     5    5     FB5N_R     FB5N   
...          ...   ...  ...   ...  ...  ...   ...  ...        ...      ...   
2506   987273073  2507    3     8    8  409   604  629  (PVL05)_L      NaN   
2507   987842109  2508    3     9   23  533   780  815        NaN      NaN   
2508   988567837  2509    2     3    4   16    58   63     FB4G_R     FB4G   
2509   988909130  2510    2     3    4  389   559  572     FB5V_R     FB5V   
2510   989228019  2511    2     3    4  389   559  572     FB5V_R     FB5V   

      ...  status  cropped      statusLabel  cellBodyFiber  somaRadius  \
0     ...  Traced    False   Roughly traced          PDM19       301.0   
1     ...  Traced    False   Roughly traced          PVL20         NaN   
2     ...  Traced    False   Roughly traced          PVL13       236.0   
3     ...  Traced    False   Roughly traced          PVL13       236.0   
4     ...  Traced    False   Roughly traced          AVM08       472.5   
...   ...     ...      ...              ...            ...         ...   
2506  ...  Traced    False   Roughly traced            NaN         NaN   
2507  ...  Orphan      NaN  Orphan hotknife            NaN         NaN   
2508  ...  Traced    False   Roughly traced          AVM08         NaN   
2509  ...  Traced    False   Roughly traced          AVM10       296.5   
2510  ...  Traced    False   Roughly traced          AVM10       296.5   

               somaLocation  \
0     [23044, 14981, 11600]   
1                       NaN   
2     [12083, 10523, 16816]   
3     [13524, 10108, 16480]   
4     [19178, 29711, 37312]   
...                     ...   
2506                    NaN   
2507                    NaN   
2508                    NaN   
2509  [13226, 32024, 18600]   
2510  [12516, 33587, 18648]   

                                                roiInfo        notes  \
0     {'INP': {'pre': 87, 'post': 351, 'downstream':...          NaN   
1     {'INP': {'pre': 464, 'post': 1327, 'downstream...          NaN   
2     {'INP': {'pre': 79, 'post': 126, 'downstream':...          NaN   
3     {'INP': {'pre': 40, 'post': 128, 'downstream':...          NaN   
4     {'SNP(L)': {'post': 5, 'upstream': 5, 'mito': ...  SMPCREFB5_4   
...                                                 ...          ...   
2506  {'SNP(R)': {'pre': 65, 'post': 52, 'downstream...          NaN   
2507  {'SNP(R)': {'pre': 2, 'post': 13, 'downstream'...          NaN   
2508  {'SNP(R)': {'pre': 6, 'post': 73, 'downstream'...  CRELALFB4_3   
2509  {'SNP(R)': {'pre': 1, 'post': 28, 'downstream'...    CRELALFB5   
2510  {'SNP(R)': {'pre': 1, 'post': 45, 'downstream'...    CRELALFB5   

                                              inputRois  \
0     ['EPA(R)', 'GOR(R)', 'IB', 'ICL(R)', 'INP', 'S...   
1     ['ATL(R)', 'IB', 'ICL(R)', 'INP', 'PLP(R)', 'S...   
2     ['ATL(R)', 'IB', 'ICL(R)', 'INP', 'SCL(R)', 'S...   
3     ['GOR(R)', 'IB', 'ICL(R)', 'INP', 'PLP(R)', 'S...   
4     ['CRE(-ROB,-RUB)(R)', 'CRE(R)', 'CX', 'FB', 'F...   
...                                                 ...   
2506  ['CRE(-ROB,-RUB)(R)', 'CRE(-RUB)(L)', 'CRE(L)'...   
2507                               ['SMP(R)', 'SNP(R)']   
2508  ['CRE(-ROB,-RUB)(R)', 'CRE(R)', 'CX', 'FB', 'F...   
2509  ['AB(R)', 'CRE(-ROB,-RUB)(R)', 'CRE(R)', 'CX',...   
2510  ['CRE(-ROB,-RUB)(R)', 'CRE(R)', 'CX', 'FB', 'F...   

                                             outputRois  
0     ['GOR(R)', 'IB', 'ICL(R)', 'INP', 'SCL(R)', 'S...  
1     ['ATL(R)', 'IB', 'ICL(R)', 'INP', 'PLP(R)', 'S...  
2     ['ATL(R)', 'IB', 'ICL(R)', 'INP', 'SCL(R)', 'S...  
3     ['I

This is the edgelist she used in June of last year along with the weights compared to the edgelist that I made in June of this year. They are clearly different.

In [32]:
# load Rhessa's data from June, it has no header
ovi_inputs_Rh = pd.read_csv('ovi_preprocessed/preprocessed_inputs-v1.2.1/only_inputs.txt', header=None, delimiter=' ')

# rename columns
ovi_inputs_Rh.columns = ['source', 'target', 'weight']
ovi_inputs_Rh

source      target  weight
0        234630133   296855409       1
1        234630133   296885362       2
2        234630133   297908595       1
3        234630133   298603730       1
4        234630133   326253554      41
...            ...         ...     ...
197132  7112622044  5813080768       3
197133  7112622044  5813096198       2
197134  7112622044  5813098223       1
197135  7112624972   576566223       1
197136  7112624972   610312734       1

[197137 rows x 3 columns]

In [33]:
# load my data from June of this year
ovi_inputs_gg = pd.read_csv('for_running_modularity/oviINr_inputs_selfremoved/original_oviINr_in_connectome_undirected_selfremoved.txt', header=None, delimiter=',')

# rename columns
ovi_inputs_gg.columns = ['source', 'target', 'weight']
ovi_inputs_gg

source      target  weight
0        234630133   296855409       4
1        234630133   296885362       2
2        234630133   297908595       1
3        234630133   298603730       1
4        234630133   326253554      41
...            ...         ...     ...
197132  7112622044  5813080768      10
197133  7112622044  5813096198       2
197134  7112622044  5813098223       5
197135  7112624972   576566223       1
197136  7112624972   610312734       1

[197137 rows x 3 columns]

In [34]:
# find the differences
differences = ovi_inputs_Rh.compare(ovi_inputs_gg)
differences

weight      
         self other
0         1.0   4.0
6         1.0   2.0
7        10.0  29.0
9         1.0   5.0
10       14.0  41.0
...       ...   ...
197120    2.0  32.0
197121    1.0   6.0
197124    1.0  11.0
197132    3.0  10.0
197134    1.0   5.0

[38090 rows x 2 columns]

In [35]:
# compare to the connectome of oviINr currently
oviINr_connectome_undirected.compare(ovi_inputs_gg)

Empty DataFrame
Columns: []
Index: []

In [36]:
# compare to the connectome of oviINr currently
oviINr_connectome_undirected.compare(ovi_inputs_Rh)

weight      
         self other
0         4.0   1.0
6         2.0   1.0
7        29.0  10.0
9         5.0   1.0
10       41.0  14.0
...       ...   ...
197120   32.0   2.0
197121    6.0   1.0
197124   11.0   1.0
197132   10.0   3.0
197134    5.0   1.0

[38090 rows x 2 columns]

The only file I have from last year that could be used to independently verify that Janelia changed the weights in June is the preprocessed edges that definitely came from Rhessa so maybe there is no point but for curiosity's sake I'll mess around with it.

In [84]:
# load edges from Rhessa's data from last August
ovi_prepro_edges = pd.read_csv('ovi_preprocessed/preprocessed_inputs-v1.2.1/preprocessed_undirected_edges.csv')
ovi_prepro_edges

Unnamed: 0  node1  node2  total_weight
0                0      1     63             1
1                1      1    161             1
2                2      1    226             1
3                3      1    281             1
4                4      1    331             3
...            ...    ...    ...           ...
197132      197132   2499   2509             1
197133      197133   2506   2510             1
197134      197134   2509   2510             3
197135      197135   2509   2511             1
197136      197136   2510   2511             8

[197137 rows x 4 columns]

In [85]:
# merge the bodyId's with the edges using the key
ovi_prepro_edges2 = ovi_prepro_edges.merge(ovi_prepro[['id',' key']], left_on='node2', right_on=' key', how='left')

# rename id column to target
ovi_prepro_edges2.rename(columns={'id': 'target'}, inplace=True)

ovi_prepro_edges2 = ovi_prepro_edges2.merge(ovi_prepro[['id',' key']], left_on='node1', right_on=' key', how='left')
ovi_prepro_edges2.rename(columns={'id': 'source'}, inplace=True)

ovi_prepro_edges2.rename(columns={'total_weight': 'weight'}, inplace=True)

In [86]:
ovi_prepro_edges2 = ovi_prepro_edges2[['source', 'target', 'weight']]

In [77]:
ovi_prepro_edges2.sort_values(by=['source', 'target'], ignore_index=True)#, inplace=True)

source     target  weight
0        234630133  296855409       1
1        234630133  296885362       2
2        234630133  297563139       3
3        234630133  297908595       1
4        234630133  298254517       1
...            ...        ...     ...
197132  7112622044  984536209       2
197133  7112622044  987117151       3
197134  7112622044  988909130      15
197135  7112622044  989228019       7
197136  7112624972  827893839       4

[197137 rows x 3 columns]